common/lib/dependabot/command_helpers.rb

```diff
@@ -0,0 +1,202 @@
+      stdout = T.let("", String)
+      stderr = T.let("", String)
+      status = T.let(nil, T.nilable(ProcessStatus))
+      pid = T.let(nil, T.untyped)
+      start_time = Time.now
+
+      begin
+        T.unsafe(Open3).popen3(*env_cmd) do |stdin, stdout_io, stderr_io, wait_thr|
+          pid = wait_thr.pid
+          stdin&.write(stdin_data) if stdin_data
+          stdin&.close
+
+          stdout_io.sync = true
+          stderr_io.sync = true
+
+          # Array to monitor both stdout and stderr
+          ios = [stdout_io, stderr_io]
+
+          last_output_time = Time.now # Track the last time output was received
+
+          until ios.empty?
+            # Calculate remaining timeout dynamically
+            remaining_timeout = timeout - (Time.now - last_output_time)
+
+            # Raise an error if timeout is exceeded
+            if remaining_timeout <= 0
+              terminate_process(pid)
+              status = ProcessStatus.new(wait_thr.value, 124)
+              raise Timeout::Error, "Timed out due to inactivity after #{timeout} seconds"
+            end
+
+            # Use IO.select with a dynamically calculated short timeout
+            ready_ios = IO.select(ios, nil, nil, [0.1, remaining_timeout].min)
+
+            # Process ready IO streams
+            ready_ios&.first&.each do |io|
+              io.set_encoding("UTF-8", "UTF-8")
+
+              data = io.read_nonblock(1024)
+
+              last_output_time = Time.now
+              if io == stdout_io
+                stdout += data
+              else
+                stderr += data unless stderr_to_stdout
+                stdout += data if stderr_to_stdout
+              end
+            rescue EOFError
+              ios.delete(io)
+            rescue IO::WaitReadable
+              next
+            end
+          end
+
+          status = ProcessStatus.new(wait_thr.value)
+        end
+      rescue Timeout::Error => e
+        stderr += e.message unless stderr_to_stdout
+        stdout += e.message if stderr_to_stdout
+      rescue Errno::ENOENT => e
+        stderr += e.message unless stderr_to_stdout
+        stdout += e.message if stderr_to_stdout
+      end
+
+      elapsed_time = Time.now - start_time
+      [stdout, stderr, status, elapsed_time]
+    end
+    # rubocop:enable Metrics/AbcSize
+    # rubocop:enable Metrics/MethodLength
+    # rubocop:enable Metrics/PerceivedComplexity
+    # rubocop:enable Metrics/CyclomaticComplexity
+
+    sig { params(pid: T.nilable(Integer)).void }
+    def self.terminate_process(pid)
+      return unless pid
+
+      begin
+        if process_alive?(pid)
+          Process.kill("TERM", pid) # Attempt graceful termination
+          sleep(0.5) # Allow process to terminate
+        end
+        if process_alive?(pid)
+          Process.kill("KILL", pid) # Forcefully kill if still running
+        end
+      rescue Errno::EPERM
+        Dependabot.logger.error("Insufficient permissions to terminate process: #{pid}")
```

Comment:
In natural way if it terminates (old way) we don't do anything. If it doesn't then the operation is going to get canceled after hanging out as how it was before. But generally we should have permission to do it since we initiated the command in this thread.

Reply:
I would be surprised if this ever triggers TBH.  This is only ever going to terminate a sub-process, and AFAIK the parent process _should_ always have perms to terminate a child process.

Reply:
I aggree.